Add F1 score, precision, and recall metrics as MultilabelSegmentation default metrics

FrenchKrab opened the pull request:
Currently the MultilabelSegmentation task has no default metric, this PR adds these 3 (torchmetrics) metrics as default.

FrenchKrab committed:
add f1,precision,recall metrics for multilabel task

FrenchKrab committed:
black formatting

hbredin commented on pyannote/audio/tasks/segmentation/multilabel.py:
```diff
@@ -265,6 +275,26 @@ def validation_step(self, batch, batch_idx: int):
+        if self.classes is not None:
+            classes = self.classes
+        else:
+            classes = self.protocol.stats()["labels"].keys()
```
Don't do that. Calling `protocol.stats()` might be veeeeeeeery slow.

Does it ever happen that `classes` is None when `default_metric` is called?
Isn't it always set in `setup(stage="fit")`?

FrenchKrab replied:
Ah ! Indeed it seems that it should never be None.

FrenchKrab committed:
simplify MultilabelSegmentation's default_metric logic

FrenchKrab committed:
better shape for tensors passed to MultilabelSegmentation metrics

FrenchKrab committed:
use macro avg for default MultilabelSegmentation metrics

FrenchKrab committed:
black format

hbredin commented on pyannote/audio/tasks/segmentation/multilabel.py:
```diff
@@ -251,10 +254,25 @@ def validation_step(self, batch, batch_idx: int):
         mask: torch.Tensor = y_true != -1
-        y_pred = y_pred[mask]
-        y_true = y_true[mask]
+        y_pred = y_pred[mask].reshape(shape)
```
This will break as soon as `mask` contains at least one `-1` because the overall size of `y_pred` will then be smaller.

FrenchKrab committed:
add support for "per class" metric for MultilabelSegmentation
Still support "global" metrics, BUT they have to be of a binary type.

FrenchKrab committed:
fix logic in multilabel setup_validation_metric

FrenchKrab committed:
rename "_per_metric" -> "_classwise"

FrenchKrab committed:
add Loggable and LoggableHistogram classes

FrenchKrab committed:
Revert "add Loggable and LoggableHistogram classes" (oops, wrong branch)

This reverts commit a890ef61ae3d43ac7b30945e83ad57524d63a876.

FrenchKrab committed:
fix wrong default_metric return value

FrenchKrab committed:
make MultilabelSegmentation global metric actually be multilabel

(uses ignore_index for that, which ignores all targets of that index for
computations, in pyannote's case : -1)

FrenchKrab committed:
update comments

FrenchKrab committed:
Merge branch 'develop' into multilabel_default_metrics

FrenchKrab committed:
small fix

FrenchKrab committed:
Merge branch 'multilabel_default_metrics' of github.com:FrenchKrab/pyannote-audio into multilabel_default_metrics

hbredin committed:
Merge branch 'develop' into multilabel_default_metrics

hbredin commented:
Is this ready for review?

FrenchKrab committed:
add ignore_index to default_metric_classwise

FrenchKrab committed:
Revert "add ignore_index to default_metric_classwise"

This reverts commit 5f51b508da9172353e75f527211387b0c92345cd.

FrenchKrab commented:
It should be now ! (although it changes more than anticipated) (sorry for the last two commits, should've reread my whole code before committing a "fix")

Multilabel metrics use ignore_index=-1 to support having targets of different labels ignored in the metrics (partially annotated data).
Classwise (binary) metrics do not need it since we can just filter out the missing data.

Maybe we should enforce ignore_index==-1 in the multilabel metrics / raise an exception, it's easy to miss and the metric would either still work (but give wrong values) or crash (and the user probably wont understand why).

hbredin commented:
As discussed right now, would be nice to try `average="none"` instead of those pesky classwise metrics :)

chai3 committed:
fix: raise TypeError on wrong device type in Pipeline.to and Inference.to

Fixes 1397

hbredin committed:
feat(task): add support for multi-task models (#1374)

BREAKING(model): get rid of (flaky) `Model.introspection`

hbredin committed:
fix(inference): fix multi-task inference

hbredin committed:
feat: update FAQtory default answer

hbredin committed:
improve(test): use pyannote.database.registry (#1413)

flyingleafe committed:
feat(pipeline): add `return_embeddings` option to `SpeakerDiarization` pipeline

Co-authored-by: Hervé BREDIN <hbredin@users.noreply.github.com>

hbredin committed:
fix: fix missed speech at the very beginning/end

hbredin committed:
doc: add note to self regarding cluster reassignment (#1419)

DiaaAj committed:
fix(doc): fix typo in diarization docstring

hbredin committed:
ci: update suggest.md (#1435)

hbredin committed:
feat: add support for WeSpeaker embeddings (#1444)

aashish-19 committed:
fix: fix security issue in FAQtory bot

hbredin committed:
Update README.md

hbredin committed:
Update README.md

FrenchKrab committed:
fix(task): fix MultiLabelSegmentation.val_monitor

hbredin committed:
Merge branch 'develop' into multilabel_default_metrics

hbredin commented:
Going over your PRs :) Is this mergeable?

hbredin committed:
fix(core): fix Model.example_output for embedding models

FrenchKrab committed:
update multilabel classwise metrics naming

FrenchKrab committed:
update multilabel default metric + docstring + black formatting

FrenchKrab commented:
I should test it before, but i'm done with the implementation (i dont know if it's ok for you though :) ).

In the end I didn't find how to do away with "metric_classwise", there's a [torchmetric classwise wrapper](https://torchmetrics.readthedocs.io/en/stable/wrappers/classwise_wrapper.html) but it doesn't do what we want.
Torchmetrics's wrapper is for metrics that return a multidim tensor, but for example the multilabel variant of [Calibration Error](https://torchmetrics.readthedocs.io/en/stable/classification/calibration_error.html#multiclasscalibrationerror) doesn't exist, while the "binary" version exists for probably every metric.

hbredin committed:
Merge branch 'develop' into multilabel_default_metrics
